subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskOptionsSupplier.java

```diff
@@ -0,0 +1,129 @@
+
+package org.gradle.api.internal.tasks;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.gradle.api.internal.tasks.options.BooleanOptionElement;
+import org.gradle.api.internal.tasks.options.BuiltInOptionElement;
+import org.gradle.api.internal.tasks.options.InstanceOptionDescriptor;
+import org.gradle.api.internal.tasks.options.OptionDescriptor;
+import org.gradle.api.internal.tasks.options.OptionElement;
+import org.gradle.api.internal.tasks.options.OptionReader;
+import org.gradle.api.specs.Specs;
+import org.gradle.internal.Pair;
+import org.gradle.util.internal.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * This class is responsible for supplying the built-in options and
+ * the {@link BooleanOptionElement mutually exclusive options} of a task,
+ * based on the task options provided by the {@link OptionReader}.
+ */
+public class TaskOptionsSupplier {
+    private static final Logger LOGGER = LoggerFactory.getLogger(TaskOptionsSupplier.class);
+
+    @VisibleForTesting
+    static final List<BuiltInOptionElement> BUILT_IN_OPTIONS = Stream.of(
+        new BuiltInOptionElement(
+            "Causes the task to be re-run even if up-to-date.",
+            "rerun",
+            task -> task.getOutputs().upToDateWhen(Specs.satisfyNone())
+        )
+    ).collect(Collectors.toList());
+
+    /**
+     * Builds a list of implicit built-in options available for every task.
+     */
+    private static List<OptionDescriptor> getBuiltInOptions(Object target, Collection<String> reserved) {
+        List<OptionDescriptor> allBuiltInOptions = BUILT_IN_OPTIONS.stream().map(optionElement ->
+            new InstanceOptionDescriptor(target, optionElement, null, reserved.contains(optionElement.getOptionName()))
+        ).collect(Collectors.toList());
+        List<OptionDescriptor> validBuiltInOptions = new ArrayList<>();
+        for (OptionDescriptor builtInOption : allBuiltInOptions) {
+            // built-in options only enabled if they do not clash with task-declared ones
+            if (builtInOption.isClashing()) {
+                LOGGER.warn("Built-in option '{}' in task {} was disabled for clashing with another option of same name", builtInOption.getName(), target);
+            } else {
+                validBuiltInOptions.add(builtInOption);
+            }
+        }
+        return validBuiltInOptions;
+    }
+
+    /**
+     * Builds a map of generated opposite options and
+     * add pairs of mutually exclusive options to the {@link TaskOptions taskOptions}.
+     */
+    private static Map<String, OptionDescriptor> getOppositeOptions(Object target, Map<String, OptionDescriptor> options, TaskOptions taskOptions) {
```

Comment:
❕ I think it's kind of unexpected behavior for a `getXYZ` method to modify an argument.  Could this be reworked to not need to do that? If not, I think it's worth calling out the modification to `taskOptions` in javadoc.

Reply:
Good point. I think it makes sense to keep the generation of these two option groups together as one is very much based on the other, but I spelt this out more clearly in the javadoc.